bundle/edu.gemini.catalog/src/main/scala/edu/gemini/catalog/votable/VoTableParser.scala

```diff
@@ -73,6 +77,34 @@ case object UCAC4Filter extends MagnitudesFilter {
   }
 }
 
+case object PPMXLFilter extends MagnitudesFilter {
+  // PPMXL may contain two representations for bands R and B, represented with ids r1mag/r2mag or b1mag/b2mac
+  // The ids r1mag/r2mag are preferred but if they are absent we should use the alternative values
+  val primaryMagnitudesIds = List("r1mag", "b1mag")
+  val alternateMagnitudesIds = List("r2mag", "b2mag")
+  val idsMapping = primaryMagnitudesIds.zip(alternateMagnitudesIds)
+
+  override def filterMagnitudeFields(magnitudeFields: List[(FieldId, Magnitude)]): List[Magnitude] = {
+    // Alternative values if primaries are not on the list or NaN
+    val alternateMagnitudes = (for {
+        m <- magnitudeFields
+        if alternateMagnitudesIds.contains(m._1.id)
+        if validMagnitude(m._2)
+      } yield (m._1.id, m)).toMap
+    // Remove the non valid magnitudes
+    val primaryMagnitudes = magnitudeFields.filter(m => validMagnitude(m._2))
+    // Collect the magnitudes present on the primary list and not on the secondary list
+    val m = primaryMagnitudes.filter {
+      case f @ (FieldId(i, _), _) if alternateMagnitudesIds.contains(i) => false
+      case _                                                            => true
+    }
+    // In case the primary isn't present use the alternative
+    idsMapping.foldLeft(m) {
+      case (mags, (id1, id2)) => if (mags.exists(_._1.id === id1)) mags else mags ++ alternateMagnitudes.get(id2)
+    }.map(_._2)
+  }
+}
```

Comment:
Could you make one pass through the `magnitudeFields` upfront to produce a `Map[String, Magnitude]` for all valid `Magnitude` values in the list.  Then do something like the `idMapping` fold at the end to fix it up?  Does the order matter?  Is this more or less the same result?:

````scala
val magMap1 = (Map.empty[String, Magnitude]/:magnitudeFields) { case (m, (FieldId(id, _), mag)) =>
   if (validMagnitude(mag)) m + (id -> mag) else m
}
// Now magMap1 might have double entries for R and B.  Just get rid of the alternative if so.
val magMap2 = (magMap1/:idsMapping) { case (m, (id1, id2)) =>
  if (magMap1.contains(id1)) m - id2 else m
}
magMap2.values.toList  // we've scrambled the order of course
````